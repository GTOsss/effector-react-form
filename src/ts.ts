import { Store, Event, Domain } from 'effector';
import React from 'react';

export type AnyState = Record<string, any>;

export type ErrorsInline = Record<string, Message>;

export type FieldsInline = Record<string, FieldState>;

export type Message = string | null | undefined;

export type Messages<Values> = {
  [key in keyof Values]?: Values[key] extends AnyState ? Messages<Values[key]> : Message;
};

export type SubmitParams<Values = any, Meta = any> = {
  values: Values;
  errorsInline: ErrorsInline;
  fieldsInline: FieldsInline;
  form: FormState;
  meta: Meta;
};

export type FormState = {
  submitted: boolean;
  hasError: boolean;
  hasOuterError: boolean;
};

export type FieldState = {
  _type: 'fieldMeta';
  active: boolean;
  touched: boolean;
  changed: boolean;
  blurred: boolean;
  touchedAfterOuterError: boolean;
  changedAfterOuterError: boolean;
  blurredAfterOuterError: boolean;
  validate?: (value: any) => string | undefined;
};

export type ControllerParams = {
  name: string;
  validate?: (value: any) => Message;
};

export type SetValueParams = {
  field: string;
  value: any;
};

export type SetOrDeleteErrorParams = {
  field: string;
  error?: Message;
};

export type SetFieldStateParams = {
  field: string;
  state: FieldState;
};

export type SetOrDeleteOuterErrorParams = {
  field: string;
  error: Message;
};

export type FieldInitParams = {
  name: string;
  validate?: ControllerParams['validate'];
};

export type ControllerInjectedResult<Meta = any> = {
  input: {
    name: string;
    value;
    onChange: (event: any) => void;
    onFocus: (event: any) => void;
    onBlur: (event: any) => void;
  };
  error: Message;
  innerError: Message;
  outerError: Message;
  isShowError: boolean;
  isShowOuterError: boolean;
  isShowInnerError: boolean;
  form: FormState;
  validate?: (value: any) => Message;
  setFieldState: ({ field: string, state: FieldState }) => void;
  setOrDeleteError: ({ field: string, error: Message }) => void;
  setOrDeleteOuterError: ({ field: string, error: Message }) => void;
  setOuterErrorsInlineState: (errors: ErrorsInline) => void;
  fieldState: FieldState;
};

export type Controller<Meta = any> = () => ControllerInjectedResult<Meta>;

export type ControllerHof<Meta = any> = (a: ControllerParams) => Controller<Meta>;

export type FormValidateParams<Values> = {
  values: Values;
  errorsInline: ErrorsInline;
};

export type FormValidate<Values> = (params: FormValidateParams<Values>) => ErrorsInline;

export type MapSubmit<Values, ResultValues, Meta = any> = (
  params: SubmitParams<Values, Meta>,
) => SubmitParams<ResultValues, Meta>;

export type OnSubmit<Values, Meta = any> = (params: SubmitParams<Values, Meta>) => void;

export type OnChange<Values, Meta = any> = OnSubmit<Values, Meta>;

export type UseErrorParams<Values> = {
  name: string;
  form: Form<Values>;
};

export type UseErrorResult<Meta = any> = {
  inputValue: any;
  form: FormState;
  fieldState: FieldState;
  error: Message;
  innerError: Message;
  outerError: Message;
  isShowError: boolean;
  isShowOuterError: boolean;
  isShowInnerError: boolean;
};

export type FieldArrayParams<Values> = {
  name: string;
  fieldArray: FieldArray<Values>;
};

export type MapFieldsArrayCallbackParams = {
  formItemName: string;
  field: any;
  fields: Array<any>;
  index: number;
};

export type MapFieldArrayCallback = (params: MapFieldsArrayCallbackParams) => React.ReactNode;

export type ResultUseFieldArray = {
  map: (fn: MapFieldArrayCallback) => React.ReactNode[];
  remove: (index: number) => void;
  push: (value: any | Array<any>) => void;
};

export type CreateFormParams<Values = any, MappedValues = any, Meta = any> = {
  validate?: FormValidate<Values>;
  mapSubmit?: MapSubmit<Values, MappedValues, Meta>;
  onSubmit?: OnSubmit<Values, Meta>;
  onChange?: OnChange<Values, Meta>;
  initialValues?: Values;
  initialMeta?: Meta;
  domain?: Domain;
};

export type Form<Values = any, Meta = any> = {
  $values: Store<Values>;
  $errorsInline: Store<ErrorsInline>;
  $outerErrorsInline: Store<ErrorsInline>;
  $form: Store<FormState>;
  $fieldsInline: Store<Record<string, FieldState>>;

  setValue: Event<any>;
  setOrDeleteError: Event<any>;
  setErrorsInlineState: Event<any>;
  setFieldState: Event<any>;
  setSubmitted: Event<any>;
  resetOuterFieldStateFlags: Event<any>;
  setOrDeleteOuterError: Event<any>;
  reset: Event<any>;

  setMeta: Event<any>;

  setOuterErrorsInlineState: Event<any>;
  validateForm: Event<any>;
  submit: Event<any>;
  onSubmit: Event<SubmitParams<Values, Meta>>;

  onChangeFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  onFocusFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  onBlurFieldBrowser: Event<{ event: React.SyntheticEvent; name: string }>;
  fieldInit: Event<{ name: string; validate?: ControllerParams['validate'] }>;
};

export type FieldArray<Values> = {
  form: Form<Values>;
  push: Event<{ fieldName: string; value: any | any[] }>;
  remove: Event<{ fieldName: string; index: number }>;
};

export type CreateFieldArrayParams<Values> = {
  form: Form<Values>;
  domain?: Domain;
};

// declare const useFieldArray: <Values extends AnyState = AnyState>(
//   paramsFieldArray: FieldArrayParams,
// ) => ResultUseFieldArray<Values>;
//
// declare const setIn: <O extends AnyState = AnyState, V = any>(
//   object: O,
//   path: string,
//   value: V,
// ) => O;
//
// declare const deleteIn: <O extends AnyState = AnyState>(
//   object: O,
//   path: string,
//   removeEmpty: boolean = false,
//   inDeep: boolean = true,
// ) => O;
//
// declare const getIn: <O extends AnyState = AnyState, V = any>(
//   object: O,
//   path: string,
//   removeEmpty: boolean = false,
//   inDeep: boolean = true,
// ) => V = any;
//
// declare const makeNested: <O extends AnyState = AnyState, R extends AnyState>(
//   inlineMap: O,
// ) => R;
//
// declare const removeFromInlineMap: <O extends FieldsInline = FieldsInline, R extends FieldsInline = FieldsInline>(
//   inlineMap: O,
//   key: string,
// ) => R;
