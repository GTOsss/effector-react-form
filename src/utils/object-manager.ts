import stringToPath from 'lodash.topath';

/**
 * @param {object} state Redux state
 * @param path
 * @param removeEmpty
 * @param inDeep should be false for inlineMap
 * @param index
 * @returns {object} State
 */
export const deleteIn = (
  state: Record<string, any>,
  path: Array<string> | string,
  removeEmpty = false,
  inDeep = true, // false for inlineMap
  index = 0,
): any => {
  let pathArray = index === 0 ? stringToPath(path) : path;
  if (!inDeep) {
    pathArray = [path];
  }

  const currentKey = pathArray[index];

  if (!state) {
    return state;
  }

  const isEndPoint = pathArray.length - 1 === index;
  const isArray = Array.isArray(state);
  const isObject = !isArray && typeof state === 'object';
  // @ts-ignore
  const newState = isArray ? [...state] : { ...state };

  if (isObject) {
    if (isEndPoint) {
      // tslint:disable-next-line:no-dynamic-delete
      delete newState[currentKey];
      return newState;
    } else {
      const result = deleteIn(newState[currentKey], pathArray, removeEmpty, inDeep, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        // tslint:disable-next-line:no-dynamic-delete
        delete newState[currentKey];
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  } else if (isArray) {
    if (isEndPoint) {
      return newState.filter((_, i) => Number(currentKey) !== i);
    } else {
      const result = deleteIn(newState[currentKey], pathArray, removeEmpty, inDeep, index + 1);
      const isRemoveEmpty = !result || (removeEmpty && (!result || !Object.keys(result).length));
      if (isRemoveEmpty) {
        return newState.filter((_, i) => Number(currentKey) !== i);
      } else {
        newState[currentKey] = result;
      }
      return newState;
    }
  }
};

/**
 *
 * @param {object} state Object will be expand
 * @param {string} path Path in object
 * @param {*} value Value will be add
 * @param {number} pathIndex Not require
 * @returns {object} State
 */
export const setIn = (state, path, value, pathIndex = 0) => {
  const pathArray = pathIndex === 0 ? stringToPath(path) : path;

  if (pathIndex >= pathArray.length) {
    return value;
  }

  const first = pathArray[pathIndex];
  const firstState = state && (Array.isArray(state) ? state[Number(first)] : state[first]);
  const next = setIn(firstState, pathArray, value, pathIndex + 1);

  if (!state) {
    if (isNaN(first)) {
      return { [first]: next };
    }
    const initialized = [];
    // @ts-ignore
    initialized[parseInt(first, 10)] = next;
    return initialized;
  }

  if (Array.isArray(state)) {
    // @ts-ignore
    const copy = [].concat(state);
    // @ts-ignore
    copy[parseInt(first, 10)] = next;
    return copy;
  }

  return {
    ...state,
    [first]: next,
  };
};

/**
 *
 * @param {object} state Redux state
 * @param {string} field Path to field
 * @param {*?} defaultValue
 * @returns {object} State
 */
export const getIn = (state, field, defaultValue?) => {
  if (!state) {
    return defaultValue;
  }

  const path = stringToPath(field);
  const { length } = path;
  if (!length) {
    return defaultValue;
  }

  let result = state;
  for (let i = 0; i < length && result; i += 1) {
    result = result[path[i]];
  }

  if (!result) {
    return defaultValue;
  }

  return result;
};

export const makeNested = <Result = Record<string, any>>(inlineMap: Record<string, any>): Result =>
  Object.entries(inlineMap).reduce((acc, [key, value]) => setIn(acc, key, value), {} as Result);

const isFieldMeta = (value) => typeof value === 'object' && value._type === 'fieldMeta';

export const removeFromInlineMap = (map: Record<string, any>, key: string) => {
  const nestedMap = deleteIn(makeNested(map), key);
  const nodes = [{ node: nestedMap, path: [] }];
  const newInlineMap = {};

  while (nodes.length) {
    const currentNode = nodes.pop();
    Object.entries(currentNode.node).forEach(([k, v]) => {
      if (isFieldMeta(v)) {
        newInlineMap[currentNode.path.concat(k).join('.')] = v;
      } else {
        nodes.push({
          node: v,
          path: currentNode.path.concat(k),
        });
      }
    });
  }

  return newInlineMap;
};

export const makeConsistentKey = (key) => stringToPath(key).join('.');
