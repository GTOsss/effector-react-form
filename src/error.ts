import {useStoreMap} from 'effector-react';
import {ErrorsInline, FieldsInline, FieldState, FormState, Message, UseErrorParams} from '../index';
import {getIn} from './utils/object-manager';
import {initialFieldState} from './form';

const useError = <Values>({
  name,
  $values,
  $errorsInline,
  $outerErrorsInline,
  $form,
  $fieldsInline,
}: UseErrorParams) => {
  const value = useStoreMap<Values, any, [string]>({
    store: $values,
    keys: [name],
    fn: (values, [field]) => getIn(values, field) || null,
  });

  const innerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $errorsInline,
    keys: [name],
    fn: (errorsInline, [field]) => errorsInline[field] || null,
  });
  const outerError = useStoreMap<ErrorsInline, Message, [string]>({
    store: $outerErrorsInline,
    keys: [name],
    fn: (outerErrorsInline, [field]) => outerErrorsInline[field] || null,
  });
  const error = innerError || outerError;

  const fieldState = useStoreMap<FieldsInline, FieldState, [string]>({
    store: $fieldsInline,
    keys: [name],
    fn: (fieldsInline, [field]) => fieldsInline[field] || initialFieldState,
  });

  const formSubmitted = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.submitted,
  });
  const formHasError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasError,
  });
  const formHasOuterError = useStoreMap<FormState, boolean, []>({
    store: $form,
    keys: [],
    fn: (formState) => formState.hasOuterError,
  });
  const formState: FormState = {
    submitted: formSubmitted,
    hasError: formHasError,
    hasOuterError: formHasOuterError,
  };

  const isShowInnerError = (formState.submitted || fieldState.blurred) && Boolean(innerError);
  const isShowOuterError = (!fieldState.changedAfterOuterError) && Boolean(outerError);
  const isShowError = isShowInnerError || isShowOuterError;

  return {
    inputValue: value,
    form: formState,
    fieldState,
    error,
    innerError,
    outerError,
    isShowError,
    isShowOuterError,
    isShowInnerError,
  };
};

export default useError;
