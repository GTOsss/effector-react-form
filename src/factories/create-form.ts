import {
  ControllerParams,
  CreateFormParams,
  ErrorsInline,
  FieldsInline,
  FieldState,
  Form,
  FormState,
  Message,
  SetOrDeleteErrorParams,
} from '../ts';
import { combine, createEvent as createEventNative, createStore as createStoreNative, sample } from 'effector';
import { initialFieldState, initialFormState } from '../default-states';
import { SyntheticEvent } from 'react';
import { getValue } from '../utils/dom-helper';
import { deleteIn, getIn, setIn } from '../utils/object-manager';

const createForm = <Values>({
  validate,
  mapSubmit,
  onSubmit,
  onChange: onChangeForm,
  initialValues,
  domain,
}: CreateFormParams<Values> = {}): Form<Values> => {
  const createEvent = domain ? domain.createEvent : createEventNative;
  const createStore = domain ? domain.createStore : createStoreNative;

  const setValue = createEvent<{ field: string; value: any }>(`hookForm_SetValue`);
  const setOrDeleteError = createEvent<SetOrDeleteErrorParams>(`hookForm_SetOrDeleteError`);
  const setErrorsInlineState = createEvent<ErrorsInline>(`hookForm_SetErrorsInlineState`);
  const setFieldState = createEvent<{ field: string; state: FieldState }>(`hookForm_SetFieldState`);
  const setSubmitted = createEvent<boolean>(`hookForm_SetSubmitted`);
  const resetOuterFieldStateFlags = createEvent('hookForm_ResetOuterFieldStateFlags');
  const setOrDeleteOuterError = createEvent<{ field: string; error: Message }>('hookForm_SetOrDeleteOuterError');

  const setOuterErrorsInlineState = createEvent<ErrorsInline>('hookForm_SetOuterErrorsInlineState');
  const validateForm = createEvent('hookForm_ValidateForm');
  const submit = createEvent('hookForm_Submit');

  const $values = createStore<Values>(initialValues || ({} as Values));
  const $errorsInline = createStore<ErrorsInline>({});
  const $outerErrorsInline = createStore<ErrorsInline>({});
  const $fieldsInline = createStore<FieldsInline>({});
  const $form = createStore<FormState>(initialFormState);

  const onChangeFieldBrowser = createEvent<{ event: SyntheticEvent; name: string }>(`hookForm_OnChange`);
  const onChangeField = onChangeFieldBrowser.map<{ value: any; name: string }>(({ name, event }) => ({
    value: getValue(event),
    name,
  }));
  const onFocusFieldBrowser = createEvent<{ event: SyntheticEvent; name: string }>(`hookForm_OnFocus`);
  const onBlurFieldBrowser = createEvent<{ event: SyntheticEvent; name: string }>(`hookForm_OnBlur`);
  const fieldInit = createEvent<{ name: string; validate?: ControllerParams['validate'] }>(`hookForm_fieldInit`);

  const validateByValues = (values) => {
    const errorsInlineState = {};

    Object.entries<FieldState>($fieldsInline.getState()).forEach(([name, { validate }]) => {
      const error = validate && validate(getIn(values, name));
      if (error) {
        errorsInlineState[name] = validate && validate(getIn(values, name));
      }
    });

    if (validate) {
      const formLevelErrorsInlineState = validate({ values, errorsInline: errorsInlineState });
      Object.entries(formLevelErrorsInlineState).forEach(([name, error]) => {
        if (error) {
          errorsInlineState[name] = error;
        } else {
          delete errorsInlineState[name];
        }
      });
    }

    return errorsInlineState;
  };

  submit.watch(() => {
    setSubmitted(true);

    validateForm();
    resetOuterFieldStateFlags();

    const submitParams = {
      values: $values.getState(),
      errorsInline: $errorsInline.getState(),
      fieldsInline: $fieldsInline.getState(),
      form: $form.getState(),
    };

    if (onSubmit) {
      onSubmit(mapSubmit ? mapSubmit(submitParams) : submitParams);
    }
  });

  sample({
    source: $values,
    clock: validateForm,
    fn: (values) => validateByValues(values),
    target: $errorsInline,
  });
  sample({
    source: $values,
    fn: (values) => validateByValues(values),
    target: $errorsInline,
  });

  $values
    .on(setValue, (state, { field, value }) => setIn(state, field, value))
    .on(onChangeField, (state, { value, name }) => setIn(state, name, value));

  $errorsInline
    .on(setOrDeleteError, (state, { field, error }) =>
      error ? { ...state, [field]: error } : deleteIn(state, field, false, false),
    )
    .on(setErrorsInlineState, (_, errorsInline) => errorsInline);

  $outerErrorsInline
    .on(setOrDeleteOuterError, (state, { field, error }) =>
      error ? { ...state, [field]: error } : deleteIn(state, field, false, false),
    )
    .on(setOuterErrorsInlineState, (_, errorsInline) => errorsInline);

  $fieldsInline
    .on(setOrDeleteOuterError, (state, { field }) => ({
      ...state,
      [field]: {
        ...state[field],
        touchedAfterOuterError: false,
        changedAfterOuterError: false,
        blurredAfterOuterError: false,
      },
    }))
    .on(resetOuterFieldStateFlags, (state) => {
      const newState = {};
      Object.entries<FieldState>(state).forEach(
        ([field, state]) =>
          (newState[field] = {
            ...state,
            touchedAfterOuterError: false,
            changedAfterOuterError: false,
            blurredAfterOuterError: false,
          }),
      );
      return newState;
    })
    .on(setFieldState, (state, { field, state: fieldState }) => {
      return { ...state, [field]: fieldState };
    })
    .on(fieldInit, (state, { name, validate }) =>
      state[name] ? state : { ...state, [name]: { ...initialFieldState, validate } },
    );

  $form
    .on($outerErrorsInline.updates, (state, outerErrors) => ({
      ...state,
      hasOuterError: Boolean(Object.keys(outerErrors).length),
    }))
    .on(setSubmitted, (state, value) => setIn(state, 'submitted', value))
    .on($errorsInline.updates, (state, errorsInline) =>
      setIn(state, 'hasError', Boolean(Object.keys(errorsInline).length)),
    );

  /// Field {

  if (onChangeForm) {
    const $allFormState = sample({
      source: {
        values: $values,
        errorsInline: $errorsInline,
        outerErrorsInline: $outerErrorsInline,
        fieldsInline: $fieldsInline,
        form: $form,
      },
      clock: onChangeFieldBrowser,
    });

    if (mapSubmit) {
      const $mappedAllFormState = $allFormState.map(mapSubmit);
      $mappedAllFormState.watch(onChangeForm);
    } else {
      $allFormState.watch(onChangeForm);
    }
  }

  sample({
    source: combine({
      fieldsInline: $fieldsInline,
      outerErrorsInline: $outerErrorsInline,
    }),
    clock: onFocusFieldBrowser,
    fn: ({ fieldsInline, outerErrorsInline }, { name }) => ({
      ...fieldsInline,
      [name]: {
        ...fieldsInline[name],
        active: true,
        touched: true,
        touchedAfterOuterError: Boolean(outerErrorsInline[name]),
      },
    }),
    target: $fieldsInline,
  });
  sample({
    source: combine({
      fieldsInline: $fieldsInline,
      outerErrorsInline: $outerErrorsInline,
    }),
    clock: onChangeFieldBrowser,
    fn: ({ fieldsInline, outerErrorsInline }, { name }) => ({
      ...fieldsInline,
      [name]: {
        ...fieldsInline[name],
        changed: true,
        changedAfterOuterError: Boolean(outerErrorsInline[name]),
      },
    }),
    target: $fieldsInline,
  });
  sample({
    source: combine({
      fieldsInline: $fieldsInline,
      outerErrorsInline: $outerErrorsInline,
    }),
    clock: onBlurFieldBrowser,
    fn: ({ fieldsInline, outerErrorsInline }, { name }) => ({
      ...fieldsInline,
      [name]: {
        ...fieldsInline[name],
        active: false,
        blurred: true,
        blurredAfterOuterError: Boolean(outerErrorsInline[name]),
      },
    }),
    target: $fieldsInline,
  });

  /// }

  return {
    setValue,
    setOrDeleteError,
    setErrorsInlineState,
    setFieldState,
    setSubmitted,
    resetOuterFieldStateFlags,
    setOrDeleteOuterError,
    setOuterErrorsInlineState,
    validateForm,
    submit,

    $values,
    $errorsInline,
    $outerErrorsInline,
    $fieldsInline,
    $form,

    onChangeFieldBrowser,
    onFocusFieldBrowser,
    onBlurFieldBrowser,
    fieldInit,
  };
};

export default createForm;
