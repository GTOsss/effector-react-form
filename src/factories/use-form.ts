import React, { useCallback, useEffect, useRef } from 'react';
import {
  ControllerHof,
  ControllerInjectedResult,
  FieldState,
  FormState,
  FormFactory,
  AnyState,
  ErrorsInline,
  FieldsInline,
  Message,
} from '../ts';
import { useStoreMap } from 'effector-react';
import { getIn, makeConsistentKey } from '../utils/object-manager';
import { initialFieldState } from '../default-states';

type UseFormParamsWithFactory<Values> = {
  form: FormFactory<Values>;
};

type UseFormResultWithFactory = {
  controller: ControllerHof;
  handleSubmit: (e: React.SyntheticEvent<HTMLFormElement>) => void;
};

const useForm = <Values extends AnyState = AnyState>({
  form,
}: UseFormParamsWithFactory<Values>): UseFormResultWithFactory => {
  const {
    $values,
    $form,
    $fieldsInline,
    $errorsInline,
    $outerErrorsInline,

    setValue,
    setOrDeleteError,
    setErrorsInlineState,
    setFieldState,
    setSubmitted,
    resetOuterFieldStateFlags,
    setOrDeleteOuterError,

    setOuterErrorsInlineState,
    validateForm,
    submit,

    onChangeFieldBrowser,
    onFocusFieldBrowser,
    onBlurFieldBrowser,
    fieldInit,
  } = form;

  useEffect(() => {
    validateForm();
  }, []);

  const controller = useCallback<ControllerHof>(({ name: nameProp, validate }) => {
    return (): ControllerInjectedResult => {
      const refName = useRef<string>(makeConsistentKey(nameProp));
      refName.current = makeConsistentKey(nameProp);

      useEffect(() => {
        fieldInit({ name: refName.current });
      }, []);

      const value = useStoreMap<Values, any, [string]>({
        store: $values,
        keys: [nameProp],
        fn: (values, [field]) => getIn(values, field) || null,
      });

      const innerError = useStoreMap<ErrorsInline, Message, [string]>({
        store: $errorsInline,
        keys: [nameProp],
        fn: (errorsInline, [field]) => errorsInline[field] || null,
      });
      const outerError = useStoreMap<ErrorsInline, Message, [string]>({
        store: $outerErrorsInline,
        keys: [nameProp],
        fn: (outerErrorsInline, [field]) => outerErrorsInline[field] || null,
      });
      const error = innerError || outerError;

      const fieldState = useStoreMap<FieldsInline, FieldState, [string]>({
        store: $fieldsInline,
        keys: [nameProp],
        fn: (fieldsInline, [field]) => fieldsInline[field] || initialFieldState,
      });

      const formSubmitted = useStoreMap<FormState, boolean, []>({
        store: $form,
        keys: [],
        fn: (formState) => formState.submitted,
      });
      const formHasError = useStoreMap<FormState, boolean, []>({
        store: $form,
        keys: [],
        fn: (formState) => formState.hasError,
      });
      const formHasOuterError = useStoreMap<FormState, boolean, []>({
        store: $form,
        keys: [],
        fn: (formState) => formState.hasOuterError,
      });
      const formState: FormState = {
        submitted: formSubmitted,
        hasError: formHasError,
        hasOuterError: formHasOuterError,
      };

      const isShowInnerError = (formState.submitted || fieldState.blurred) && Boolean(innerError);
      const isShowOuterError = !fieldState.changedAfterOuterError && Boolean(outerError);
      const isShowError = isShowInnerError || isShowOuterError;

      return {
        input: {
          name: nameProp,
          value,
          onChange: (eventOrValue) => onChangeFieldBrowser({ event: eventOrValue, name: refName.current }),
          onFocus: (event) => onFocusFieldBrowser({ event, name: refName.current }),
          onBlur: (event) => onBlurFieldBrowser({ event, name: refName.current }),
        },
        form: formState,
        fieldState,
        error,
        innerError,
        outerError,
        isShowError,
        isShowOuterError,
        isShowInnerError,
        validate,
        setFieldState,
        setOrDeleteError,
        setOrDeleteOuterError,
        setOuterErrorsInlineState,
      };
    };
  }, []);

  const handleSubmit = useCallback((e: React.SyntheticEvent<HTMLFormElement>): void => {
    e.preventDefault();
    submit();
  }, []);

  return {
    controller,
    handleSubmit,
  };
};

export default useForm;
