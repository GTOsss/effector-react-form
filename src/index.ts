import useForm from './use-form';
import useFieldArray from './use-field-array';
import useError from './use-error';
import createFieldArray from './factories/create-field-array';
import createForm from './factories/create-form';
import { deleteIn, getIn, setIn, makeNested, removeFromInlineMap } from './utils/object-manager';

export * from './ts';
export {
  createForm,
  createFieldArray,
  useForm,
  useError,
  useFieldArray,
  deleteIn,
  getIn,
  setIn,
  makeNested,
  removeFromInlineMap,
};
