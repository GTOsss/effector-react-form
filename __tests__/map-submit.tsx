import React from 'react';
import { createStore } from 'effector';
import { render, fireEvent, screen } from '@testing-library/react';
import { useForm } from '../src';
import { getIn } from '../src/utils/object-manager';
import { createForm } from '../src/factories/create-form';
import { MapSubmit, Controller } from '../src/ts';

interface MappedValues {
  profile: {
    firstName?: string;
  };
  info: {
    username?: string;
  };
}

interface Values {
  username?: string;
  profile?: {
    firstName?: string;
  };
}

// @ts-ignore
const mapSubmit: MapSubmit<Values, MappedValues> = ({ values, ...rest }) => {
  const currentValues = {
    profile: { firstName: getIn(values, 'profile.firstName') },
    info: { username: getIn(values, 'username') },
  };

  return { ...rest, values: currentValues };
};

interface InputProps {
  controller: Controller;
  label: string;
}

const renderForm = ({ form }) => {
  const $fieldsInline = createStore({});

  const Input: React.FC<InputProps> = ({ controller, label }) => {
    const { input, error, isShowError } = controller();
    return (
      <div role="wrapper-for-input" className="input-wrap">
        <label>{label}</label>
        <input {...input} value={input.value || ''} role="textbox" className="input" placeholder={label} />
        {isShowError && <span>{error}</span>}
      </div>
    );
  };

  const SimpleForm = () => {
    const { handleSubmit, controller } = useForm({ form });

    return (
      <form onSubmit={handleSubmit}>
        <Input label="Username" controller={controller({ name: 'username' })} />
        <Input label="First name" controller={controller({ name: 'profile.firstName' })} />
        <button type="submit">submit</button>
      </form>
    );
  };

  render(<SimpleForm />);

  return { $fieldsInline };
};

describe('MapSubmit', () => {
  test('onSubmit form without mapSubmit', () => {
    const onSubmit = jest.fn(() => null);
    const form = createForm({ onSubmit });
    renderForm({ form });
    fireEvent.change(screen.getByPlaceholderText('Username'), { target: { value: 'test username' } });
    fireEvent.change(screen.getByPlaceholderText('First name'), { target: { value: 'test first name' } });
    const buttonSubmit = screen.getByText('submit');
    fireEvent.click(buttonSubmit);
    // @ts-ignore
    expect(onSubmit.mock.calls[0][0]).toMatchSnapshot();
  });

  test('onSubmit form with mapSubmit', () => {
    const onSubmit = jest.fn(() => null);
    const form = createForm({ onSubmit, mapSubmit });
    renderForm({ form });
    fireEvent.change(screen.getByPlaceholderText('Username'), { target: { value: 'test username' } });
    fireEvent.change(screen.getByPlaceholderText('First name'), { target: { value: 'test first name' } });
    const buttonSubmit = screen.getByText('submit');
    fireEvent.click(buttonSubmit);
    // @ts-ignore
    expect(onSubmit.mock.calls[0][0]).toMatchSnapshot();
  });

  test('onChange form without mapSubmit', () => {
    const onChange = jest.fn(() => null);
    const form = createForm({ onChange });
    renderForm({ form });
    fireEvent.change(screen.getByPlaceholderText('Username'), { target: { value: 'test username' } });
    fireEvent.change(screen.getByPlaceholderText('First name'), { target: { value: 'test first name' } });
    // @ts-ignore
    expect(onChange.mock.calls[1][0]).toMatchSnapshot();
  });

  test('onChange form with mapSubmit', () => {
    const onChange = jest.fn(() => null);
    const form = createForm({ onChange, mapSubmit });
    renderForm({ form });
    fireEvent.change(screen.getByPlaceholderText('Username'), { target: { value: 'test username' } });
    fireEvent.change(screen.getByPlaceholderText('First name'), { target: { value: 'test first name' } });
    // @ts-ignore
    expect(onChange.mock.calls[1][0]).toMatchSnapshot();
  });
});
