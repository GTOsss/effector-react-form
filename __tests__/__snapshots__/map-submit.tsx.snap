// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapSubmit onChange form with mapSubmit 1`] = `
Object {
  "errorsInline": Object {},
  "fieldsInline": Object {
    "profile.firstName": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": false,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
    "username": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
  },
  "form": Object {
    "hasError": false,
    "hasOuterError": false,
    "submitted": false,
  },
  "meta": Object {},
  "outerErrorsInline": Object {},
  "values": Object {
    "info": Object {
      "username": "test username",
    },
    "profile": Object {
      "firstName": "test first name",
    },
  },
}
`;

exports[`MapSubmit onChange form without mapSubmit 1`] = `
Object {
  "errorsInline": Object {},
  "fieldsInline": Object {
    "profile.firstName": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": false,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
    "username": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
  },
  "form": Object {
    "hasError": false,
    "hasOuterError": false,
    "submitted": false,
  },
  "meta": Object {},
  "outerErrorsInline": Object {},
  "values": Object {
    "profile": Object {
      "firstName": "test first name",
    },
    "username": "test username",
  },
}
`;

exports[`MapSubmit onSubmit form with mapSubmit 1`] = `
Object {
  "errorsInline": Object {},
  "fieldsInline": Object {
    "profile.firstName": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
    "username": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
  },
  "form": Object {
    "hasError": false,
    "hasOuterError": false,
    "submitted": true,
  },
  "meta": Object {},
  "outerErrorsInline": Object {},
  "values": Object {
    "info": Object {
      "username": "test username",
    },
    "profile": Object {
      "firstName": "test first name",
    },
  },
}
`;

exports[`MapSubmit onSubmit form without mapSubmit 1`] = `
Object {
  "errorsInline": Object {},
  "fieldsInline": Object {
    "profile.firstName": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
    "username": Object {
      "_type": "fieldMeta",
      "active": false,
      "blurred": false,
      "blurredAfterOuterError": false,
      "changed": true,
      "changedAfterOuterError": false,
      "touched": false,
      "touchedAfterOuterError": false,
      "validate": undefined,
    },
  },
  "form": Object {
    "hasError": false,
    "hasOuterError": false,
    "submitted": true,
  },
  "meta": Object {},
  "outerErrorsInline": Object {},
  "values": Object {
    "profile": Object {
      "firstName": "test first name",
    },
    "username": "test username",
  },
}
`;
